### PR TITLE
Per block transition and attestation

### PR DESCRIPTION
**Pull request by hwwhww**
Apologize for the fat PR. 🙈 

### What was wrong?
#1381 - StateMachine - per block transition and attestation

#### Follow [the beacon chain spec - Per-block processing section](https://github.com/ethereum/eth2.0-specs/blob/fe74c7e299ff90d56f4099733754c0e94e645ccc/specs/beacon-chain.md#per-block-processing) **except**:
1. No `update_ancestor_hashes`  design in this PR
2. No RANDAO reveal design in this PR
3. No versioning design in this PR

#### Modifications:
1. `helpers.py`
	1. Bugfix `get_hashes_to_sign` - shouldn’t re-hash the latest hash
	2. Remove multiple slots per committee option: https://github.com/ethereum/eth2.0-specs/pull/62
	3. Rename `get_proposer_position` to `get_block_committees_info`, it will return `BlockCommitteesInfo` as a simple structure of the committee information context.
2. `eth.beacon.state_machines.base.py`
	1. **Import block APIs**
		1. Import the given block into state machine
		2. Call process APIs; if valid, update the states in memory
	2. **Process block APIs** (all are classmethods)
		1. Pure functions for processing a block
		2. `compute_per_block_transtion`: compute for every block, will return the updated `active_state`. 
			1. Validate the `validate_parent_block_proposer` and `attestations`
			2. Return the new `active_state`
		3. (TODO in other PRs) `compute_cycle_transitions`: compute the cycle transitions. (`CYCLE_LENGTH = 64 slots`), will return the updated `crystallized_state` and `active_state`.
	3. **Proposer APIs** 
		1.  Proposer story, please see `test_propose_block_and_validate_attestation` tests:
			 1. The block proposer of slot 1:
				 1. Creates a block and process the block `B1`.
					 * Note: no shard block validation currently, so I just set `shard_block_hash=ZERO_HASH32`.
				 2. Attests the block and create an `AttestationRecord` of `B1`
				 3. Broadcasts `B1` **and the attestation record of  `B1`.**
			 2. The other attesters validate `B1`.
				 * If it's valid, sign the same message.
			 3. The next block proposer or other nodes collect  the votes
				* Aggregate the signatures and build attestation record.
				* Broadcast the attestation record.
			 4. The block proposer of slot 2:
				 1. Create a block `B2` with the aggregated attestation records
				 2. Attests the block and create an `AttestationRecord`
				 3. Broadcasts `B2` **and the attestation record of  `B2`.**
                          (Continue)
	4. **Validation APIs** (all are classmethods)
		1. proposer validation 
		2. attestation validation

#### TODO
- Add tests for validation APIs
- Add tests for aggregate signature
- Add test for multiple attestations and multiple blocks.
- Add assertions in the current tests.
- Too many arguments passing! 
	* Yep I see > 6 is a bad signal. I’d prefer to keep using pure function in processing the block, so need to refactor it.
	* FYI: The only reason why Proposer APIs are not classmethods is because it calls `self.get_attestation_record_class()`, any other way to avoid it?

@djrtwo: the proposer logic is partially from old `mock_make_attestations` and partially something new, could you verify the description above? Thanks!

#### Cute Animal Picture
Cat-sper!
caspercat_by_steve_begin

### CODE REVIEW

**Comment by djrtwo** [comment]
@pipermerriam I'm going to review this PR from a spec validity perspective

File: eth/beacon/state_machines/base.py [hunk 1]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+
+        The proposer broadcasts zir attestation with the proposed block.
+        """
+        block_committees_info = get_block_committees_info(
+            block,
+            post_crystallized_state,
+            cycle_length,
+        )
+        # Vote
+        attester_bitfield = set_voted(
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: Not for this PR, but it might be nice to create a class to represent the voting bitfield

```python
class VoteBitfield(bytes):
    @classmethod
    def create_empty(cls, size):
        ...

    def set_voted(self, idx):
        ...

    def get_vote(self, idx):
        ...
```

I don't necesarrily think those are the best APIs, just meant to be illustrative.

API for interacting with it could be similar to the `Receipt.bloom_filter`.  Also, might be nice for the class to be immutable so that each method call returns the updated bitfield, allowing you to chain multiple calls.

File: eth/beacon/state_machines/base.py [hunk 2]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+            get_empty_bitfield(block_committees_info.proposer_committee_size),
+            block_committees_info.proposer_index_in_committee,
+        )
+
+        # Get justified_slot and justified_block_hash
+        justified_slot = post_crystallized_state.last_justified_slot
+        justified_block_hash = chaindb.get_canonical_block_by_slot(justified_slot).hash
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: It may be nice to have both `get_canonical_block_hash_by_slot` and `get_canonical_block_by_slot` so that you can save the extra overhead of instantiating the block and then hashing it.  It's likely that the implementation already has to lookup the hash in order to lookup the block body.

File: eth/beacon/state_machines/base.py [hunk 3]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+            block.shard_block_hash,
+            crystallized_state.last_justified_slot,
+        )
+        # Update the bitfield and append the signatures
+        bitfield = proposer_attestation.bitfield
+        sigs = []
+        for (committee_index, sig, public_key) in votes:
+            if bls.verify(message, public_key, sig):
+                bitfield = set_voted(bitfield, committee_index)
+                sigs.append(sig)
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: Just looking at what this looks like without the mutation

```python
sigs_with_committe_info = tuple(
    sig, committee_index
    for (committee_index, sig, public_key)
    in votes
    if bls.verify(message, public_key, sig)
)
sigs, commitee_indices = zip(*sigs_with_committe_info)   # may need to handle edge case of empty list
bitfield = pipe(
    proposer_attestation.bitfield
    *(functools.partial(set_voted, index=committee_index) for committee_index in committee_indices)
)
```

That's all psuedo code, probably some minor errors, but it removes the mutation and seems readable enough).  The `functools.partial` can be removed if `set_voted` was made curriable.

File: eth/beacon/state_machines/base.py [hunk 4]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+        if block.slot_number == 0:
+            return
+
+        block_committees_info = get_block_committees_info(
+            parent_block,
+            crystallized_state,
+            cycle_length,
+        )
+
+        if len(block.attestations) == 0:
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: Rather than pre-check, I think the more idiomatic way to do this is:

```python
try:
    attestation = block.attestations[0]
except IndexError:
    raise ValidationError(...)
```

File: eth/beacon/state_machines/base.py [hunk 5]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+        )
+        cls.validate_aggregate_sig(
+            crystallized_state,
+            attestation,
+            attestation_indices,
+            parent_hashes,
+        )
+
+    @classmethod
+    def validate_slot(cls,
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: This looks like it can be a staticmethod which also suggests it could be extracted to a standalone to simplify the class (but still tacked onto the class as a `staticmethod` for convenience if you want).

File: eth/beacon/state_machines/base.py [hunk 6]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+                "Attestation slot number too low:\n"
+                "\tFound: %s, Needed greater than or equalt to: %s" %
+                (
+                    attestation.slot,
+                    max(parent_block.slot_number - cycle_length + 1, 0)
+                )
+            )
+
+    @classmethod
+    def validate_justified(cls,
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: Same about `staticmethod`, extraction to standalone.

File: eth/beacon/state_machines/base.py [hunk 7]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+                "justified_block_hash %s is not in the canonical chain" %
+                attestation.justified_block_hash
+            )
+        if justified_block.slot_number != attestation.justified_slot:
+            raise ValidationError(
+                "justified_slot %s doesn't match justified_block_hash" % attestation.justified_slot
+            )
+
+    @classmethod
+    def validate_bitfield(cls,
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: Same about `staticmethod`, extraction to standalone.

File: eth/beacon/state_machines/base.py [hunk 8]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+        Return the :class:`~eth.beacon.types.attestation_records.AttestationRecord` class that this
+        StateMachine uses for the current fork version.
+        """
+        if cls.attestation_record_class is None:
+            raise AttributeError("No `attestation_record_class` has been set for this StateMachine")
+        else:
+            return cls.attestation_record_class
+
+    @classmethod
+    def create_signing_message(cls,
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: This looks like it could be extracted to a standalone function (which will simplify this class which is quite large).  Or, at minimum, it can be made into a `staticmethod`

File: eth/beacon/state_machines/base.py [hunk 9]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+
+        # check if end bits are zero
+        last_bit = len(attestation_indices)
+        if last_bit % 8 != 0:
+            for i in range(8 - last_bit % 8):
+                if has_voted(attestation.attester_bitfield, last_bit + i):
+                    raise ValidationError("Attestation has non-zero trailing bits")
+
+    @classmethod
+    def validate_aggregate_sig(cls,
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: same about `staticmethod` -> extraction to standalone (since `create_signing_message` is also able to be made standalone.

File: eth/beacon/state_machines/base.py [hunk 10]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+        )
+        if not bls.verify(message, bls.aggregate_pubs(pub_keys), attestation.aggregate_sig):
+            raise ValidationError("Attestation aggregate signature fails")
+
+    @classmethod
+    def validate_version(cls,
+                         crystallized_state: CrystallizedState,
+                         attestation: 'AttestationRecord') -> None:
+        # TODO: it's a stub
+        return
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: *nitpick*, normally `pass` instead of `return`

File: eth/beacon/state_machines/base.py [hunk 11]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+
+    #
+    # State roots validation
+    #
+    @classmethod
+    def validate_state_roots(cls,
```
Comment (pipermerriam) [inline eth/beacon/state_machines/base.py]: Same about extraction to standalone.

File: tests/beacon/state_machines/test_proposer.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+from eth.beacon.helpers import (
+    get_block_committees_info,
+    get_new_recent_block_hashes,
+)
+
+
+@pytest.fixture
+def fixture_sm_class(config):
+    return SerenityBeaconStateMachine.configure(
```
Comment (pipermerriam) [inline tests/beacon/state_machines/test_proposer.py]: dropping the `Beacon` from the class name would make this a bit less of a mouthful to read: `SerenityStateMachine`

fits with the EVM naming convention since we don't name things `FrontierEVMVM` or `FrontierChainVM`

File: tests/beacon/state_machines/test_proposer.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        sm.crystallized_state,
+        sm.active_state,
+        block_1_shell,
+        sm.chaindb,
+        sm.config,
+    )
+    block_1 = block_1_shell.copy(
+        active_state_root=active_state_1.hash,
+    )
+    sm.import_block(block_1)
```
Comment (pipermerriam) [inline tests/beacon/state_machines/test_proposer.py]: I suspect this still has value as a smoke test but surprised to see this test has no assertions.

File: tests/beacon/state_machines/test_proposer.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+@pytest.mark.parametrize(
+    (
+        'num_validators,'
+        'cycle_length,min_committee_size,shard_count'
+    ),
+    [
+        (1000, 20, 10, 100),
+    ]
+)
+def test_propose_block_and_validate_attestation(fixture_sm_class,
```
Comment (pipermerriam) [inline tests/beacon/state_machines/test_proposer.py]: Similar lack of assertions (though I recognize that the `validate_` calls are effectively assertions.

File: eth/beacon/state_machines/base.py [hunk 15]
```diff
@@ -55,9 +76,17 @@ class BeaconStateMachine(BaseBeaconStateMachine):
     _crytallized_state = None  # type: CrystallizedState
     _active_state = None  # type: ActiveState
 
-    def __init__(self, chaindb: BaseBeaconChainDB, block: BaseBeaconBlock) -> None:
+    def __init__(self, chaindb: BaseBeaconChainDB, block: BaseBeaconBlock=None) -> None:
         self.chaindb = chaindb
-        self.block = self.get_block_class().from_block(block)
+        if self.block is None:
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/state_machines/base.py]: Sorry, it's typo. Should be: `if block is None:`

File: eth/beacon/state_machines/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/state_machines/base.py]: It came from the spec. I think it's kind of a convention in cryptography when we call the thing we sign for? Because in here, the thing we sign with BLS *is* the whole`blake(...)` part.

File: eth/beacon/state_machines/base.py [hunk 16]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+                                    chaindb: BaseBeaconChainDB,
+                                    cycle_length: int) -> ActiveState:
+        """
+        Process ``block`` and return the new ActiveState.
+
+
+        TODO: It doesn't match the latest spec.
+        There will be more fields need to be updated in ActiveState.
+        """
+        parent_block = chaindb.get_block_by_hash(block.parent_hash)
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/state_machines/base.py]: @pipermerriam I left some TODOs hints in [`BeaconChainDB`](https://github.com/ethereum/py-evm/blob/96fc4ede33378823b470f6540245b1fc94752d1d/eth/beacon/db/chain.py#L635-L650). Am I correct that since All PoW chain forks use the same block fields format, so currently EVM doesn't handle it as well?

File: eth/beacon/state_machines/base.py [hunk 17]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+    def propose_block(
+        self,
+        crystallized_state: CrystallizedState,
+        active_state: ActiveState,
+        block: BaseBeaconBlock,
+        shard_id: int,
+        shard_block_hash: Hash32,
+        chaindb: BaseBeaconBlock,
+        config: BeaconConfig,
+        private_key: int
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/state_machines/base.py]: Yes, the key pairing is in ECC family. But the curve will need to be upgraded; right now we are using `BN128` in the codebase, but will have to change to [`BLS12-381`](https://z.cash/blog/new-snark-curve/).

Boneh–Lynn–Shacham (BLS) is for aggregating the signatures into one aggregated signature, and we only have to verify the aggregated signature with public keys instead of putting every single sigs on chain.

File: eth/beacon/state_machines/base.py [hunk 18]
```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
...
+                              block: BaseBeaconBlock,
+                              shard_id: int,
+                              shard_block_hash: Hash32,
+                              chaindb: BaseBeaconBlock,
+                              cycle_length: int,
+                              private_key: int) -> 'AttestationRecord':
+        """
+        Return the initial attestation by the block proposer.
+
+        The proposer broadcasts zir attestation with the proposed block.
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/state_machines/base.py]: Ah.. was the gender-neutral possessive... but I can change to `their`.

**Comment by hwwhww** [comment]
### Changelog
1. Fix a lot PR feedback! Thanks @pipermerriam ✨ 
2. Reorg some state machine fixtures.
3. Move most of validation functions into `eth.beacon.state_machines.validation`. Only keep staticmethod `validate_attestation` for overviewing in StateMachine. (Or I should move all?) 
4. Move `create_signing_message` and `aggregate_attestation_record` to `helper.py`.
5. `aggregate_attestation_record` -> two steps (i)  `aggregate_attestation_record` and  (ii) `aggregate_votes`.
6. Add test for `aggregate_votes`. Also found the paring is too slow to add more `hypothesis` case… :|
7. Add assertions to the tests.

File: eth/beacon/block_committees_info.py [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+
+
+BlockCommitteesInfo = NamedTuple(
+    'BlockCommitteesInfo',
+    (
+        ('proposer_index', int),
+        ('proposer_index_in_committee', int),
+        ('proposer_shard_id', int),
+        ('proposer_committee_size', int),
+        ('shards_and_committees', Iterable['ShardAndCommittee'])
```
Comment (pipermerriam) [inline eth/beacon/block_committees_info.py]: Iterable is probably not what you want since that would allow a generator.  I'd actually recommend being very specific here and choosing `Tuple`.

File: eth/beacon/helpers.py [hunk 20]
```diff
@@ -360,20 +366,117 @@ def get_proposer_position(parent_block: 'BaseBeaconBlock',
...
+    try:
+        sigs, committee_indices = zip(*sigs_with_committe_info)
+    except ValueError:
+        sigs = ()
+        committee_indices = ()
+
+    sigs = sigs + tuple(pre_sigs)
+    bitfield = pre_bitfield
+    bitfield = pipe(
+        bitfield,
```
Comment (pipermerriam) [inline eth/beacon/helpers.py]: You *should* be able to just use `pre_bitfield` here.

File: tests/beacon/test_helpers.py [hunk 21]
```diff
@@ -376,16 +381,60 @@ def mock_get_shards_and_committees_for_slot(parent_block,
...
+        (5),
+        (9),
+    ],
+)
+def test_aggregate_votes(votes_count, privkeys, pubkeys):
+    bit_count = 10
+    pre_bitfield = get_empty_bitfield(bit_count)
+    pre_sigs = ()
+
+    random_votes = random.sample([i for i in range(bit_count)], votes_count)
```
Comment (pipermerriam) [inline tests/beacon/test_helpers.py]: We've *typically* tried to use `hypothesis` when we need randomness in tests so that the randomness can be reproduced upon failure.  It's not a hard requirement, but in cases where you are using random inputs to test something and the randomness can be a source of failure, it's probably what you should do.

You can do as far as having hypothesis generate values for you, or if you still just want to use the built-in python `random` standard library module, you can use this https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.random_module

File: eth/beacon/block_committees_info.py (hunk 19 above)
Comment (cburgdorf) [reply eth/beacon/block_committees_info.py]: :+1: to what @pipermerriam said.

Taking the chance to elaborate on my comment here: https://github.com/ethereum/py-evm/pull/1373#discussion_r224106008

I think as a rule of thumb: APIs *taking parmeters* (e.g. `def do_something(things: Iterable[Foo])` should accept the most flexible interface that the API can work with. I can hardly think of a reason why an API should constrain a parameter to a concrete type (e.g. `Tuple[Foo]`) if there is nothing intrinsically preventing that API from working with an `Iterable[Foo]`.

However, **Return types** should *usually* be concrete. Because, most of the time an API returning *something*, does have intrinsic knowledge about the type anyway and there is nothing to gain from making the return type more abstract if that return type fulfills that abstract interface anyway.

E.g. There is nothing to gain from doing this:

```python
def get_foo() -> Iterable[Foo]
    return [foo, too]
```

Even though that works, we are not improving flexibility here. If we make the return type `List[Foo]`, that `List[Foo]` still fulfills being `Iterable[Foo]` as well as `Sequence[Foo]` implicitly, so by making the return type somthing like `Iterable[Foo]` when in reality we have a `List[Foo]` is in fact only limiting the flexibility here.

Coming back to the code in question `shards_and_committees` is, after all, just a property on `BlocksCommitteesInfo`, hence making it fall in the return type bucket.

Related: [Covariance and Contravariance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science))

Comment (carver) [reply eth/beacon/block_committees_info.py]: Also, here is an alternative syntax, if you like it better:

```py
class BlockCommitteesInfo(NamedTuple):
  proposer_index: int
  proposer_index_in_committee: int
  # etc...
```

Comment (cburgdorf) [reply eth/beacon/block_committees_info.py]: @carver I *think* that syntax does not work with Python 3.5 though

Comment (cburgdorf) [reply eth/beacon/block_committees_info.py]: I'd be in support to drop 3.5 at any time 🤭😅

Comment (pipermerriam) [reply eth/beacon/block_committees_info.py]: Two things,

- Can't drop 3.5 until we do it in web3.py  (is v5 py36+?)
- @cburgdorf would you mind migrating that writup to https://github.com/pipermerriam/ethereum-dev-tactical-manual, maybe under a new section for static type hinting.

Comment (carver) [reply eth/beacon/block_committees_info.py]: #1391 for discussion

**Comment by hwwhww** [comment]
### Changelog
1. Fix `test_aggregate_votes` test and add `verify_votes` function
2. Add `BlockProposal` NamedTuple for having less arguments
3. Move `validate_attestation` to `eth/beacon/state_machines/validation.py`
4. Fix `validate_attestation` (should use updated `recent_block_hashes`) and add tests for every edge cases
5. Propose one more block (`block_2`) in `test_propose_block` - many lines are repeated, will consider have `evm.tools.builder.chain` tool in other PRs! (maybe after we have cycle transitions?)

File: eth/beacon/state_machines/base.py [hunk 22]
```diff
@@ -55,9 +76,17 @@ class BeaconStateMachine(BaseBeaconStateMachine):
...
     _active_state = None  # type: ActiveState
 
-    def __init__(self, chaindb: BaseBeaconChainDB, block: BaseBeaconBlock) -> None:
+    def __init__(self, chaindb: BaseBeaconChainDB, block: BaseBeaconBlock=None) -> None:
         self.chaindb = chaindb
-        self.block = self.get_block_class().from_block(block)
+        if block is None:
+            # Build a child block of current head
+            head_block = self.chaindb.get_canonical_head()
+            self.block = self.get_block_class()(*head_block).copy(
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/state_machines/base.py]: Right now maybe just in testing.

File: eth/beacon/helpers.py [hunk 23]
```diff
@@ -360,20 +366,133 @@ def get_proposer_position(parent_block: 'BaseBeaconBlock',
...
+        block.slot_number,
+        parent_hashes,
+        proposer_attestation.shard_id,
+        proposer_attestation.shard_block_hash,
+        last_justified_slot,
+    )
+
+    # Verify
+    # TODO: Verification is very slow, needs to compute in parallel
+    voting_sigs, committee_indices = verify_votes(message, votes)
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/helpers.py]: > What is going to be calling this function? This is for a block producer bundling attestations of the previous block? Note that they should be including attestations not just from the previous block but for any within the valid range (defined by parent block and CYCLE_LENGTH) that have not been included

> Note that they should be including attestations not just from the previous block but for any within the valid range (defined by parent block and CYCLE_LENGTH)

You're right. Verifying the attestations sigs *right after receiving it* makes more sense. (like an AttestationPool)
So upon receiving the Attestation:
1. Call `validate_attestation`
2. Store the attestation in the AttestationPool

When the client wants to aggregate and  the attestation pack: 
1. Check if we can reuse attestations from other block and set `oblique_parent_hashes`.
2. Aggregate the attestations that are siging for the same message.

Anything else I missed?

> We can aggregate and include attestations that are not the same as the proposer_attestation. For example, we can include votes that were for a short range fork rather than for the what the proposer submit as the head. Or some of the votes could vote for the proposer's block but for a different shard_block_hash

Right. I will rename `proposer_attestation` to something more general.

[the remark this answers is not in the document]
Comment (hwwhww) [reply eth/beacon/helpers.py]: > Right. I will rename proposer_attestation to something more general.

@djrtwo 

Second thought - I will just remove `aggregate_attestation_record` from here. It's not part of StateMachine. 😆 

But please check the correctness of the story above. Thanks!

File: eth/beacon/block_committees_info.py (hunk 19 above)
Comment (cburgdorf) [reply eth/beacon/block_committees_info.py]: @pipermerriam proposed section on typing (choosing types for function parameters and return values)! https://github.com/pipermerriam/ethereum-dev-tactical-manual/pull/24

File: eth/beacon/helpers.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (djrtwo) [reply eth/beacon/helpers.py]: Looks good. The AttestationPool and who should be doing the aggregation might change depending on the network-level protocol we define for aggregation.

I would note that the AttestationPool should likely only include attestations of single signers (non-aggregated). Also that when we receive a new block (or create one) we should clean up our attestation pool, removing any of the attestations that were included in the block.

> Check if we can reuse attestations from other block
I would be clear that this is "reuse attestations from a block on a short range fork"

> Aggregate the attestations that are signing for the same message.
And can be validly included in the block (not too old, etc)

[the remark this answers is not in the document]
Comment (djrtwo) [reply eth/beacon/helpers.py]: Right, the aggregation is more of the Node or Validator's responsibility.

File: eth/beacon/aggregation.py [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+    try:
+        sigs, committee_indices = zip(*sigs_with_committe_info)
+    except ValueError:
+        sigs = tuple()
+        committee_indices = tuple()
+
+    return sigs, committee_indices
+
+
+def aggregate_votes(bitfield: bytes,
```
Comment (djrtwo) [inline eth/beacon/aggregation.py]: I think it might be more natural to pass in an Iterable of `AttestationRecords` and return an `AttestationRecord`.

Attestations will generally be passed around and deserialized as attestation records rather than deconstructed tuples of sigs and indices

File: eth/beacon/state_machines/base.py [hunk 25]
```diff
@@ -193,9 +224,272 @@ def active_state(self) -> ActiveState:
...
+            crystallized_state: CrystallizedState,
+            active_state: ActiveState,
+            block: BaseBeaconBlock,
+            chaindb: BaseBeaconChainDB,
+            config: BeaconConfig) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState]:
+        """
+        Process ``block`` and return the new crystallized state and active state.
+        """
+        # Process per block state changes (ActiveState)
+        processing_active_state = cls.compute_per_block_transtion(
```
Comment (djrtwo) [inline eth/beacon/state_machines/base.py]: Typo
`compute_per_block_transtion` -> `compute_per_block_transition`

File: eth/beacon/state_machines/base.py [hunk 26]
```diff
@@ -193,9 +224,272 @@ def active_state(self) -> ActiveState:
...
+            validate_attestation(
+                block,
+                parent_block,
+                crystallized_state,
+                recent_block_hashes,
+                attestation,
+                chaindb,
+                cycle_length,
+            )
+
```
Comment (djrtwo) [inline eth/beacon/state_machines/base.py]: TODO: validate `block.specials` and append `active_state.pending_specials`

File: eth/beacon/state_machines/base.py [hunk 27]
```diff
@@ -193,9 +224,272 @@ def active_state(self) -> ActiveState:
...
+        active_state: ActiveState,
+        block_proposal: 'BlockProposal',
+        chaindb: BaseBeaconChainDB,
+        config: BeaconConfig,
+        private_key: int
+    ) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState, 'AttestationRecord']:
+        """
+        Propose the given block.
+        """
+        block, post_crystallized_state, post_active_state = self.process_block(
```
Comment (djrtwo) [inline eth/beacon/state_machines/base.py]: If `process_block` is to be used for calculating the state roots for a block proposal, then it needs to have the option of being run without processing the signatures from the attestations. maybe another arg -- `validate_signatures`

We can assume the attestations being added were already checked as valid when initially received/created. No need to recheck here!

File: eth/beacon/state_machines/base.py [hunk 28]
```diff
@@ -193,9 +224,272 @@ def active_state(self) -> ActiveState:
...
+            block_proposal.shard_block_hash,
+            justified_slot,
+        )
+        sigs = [
+            bls.sign(
+                message,
+                private_key,
+            )
+        ]
+        aggregate_sig = bls.aggregate_sigs(sigs)
```
Comment (djrtwo) [inline eth/beacon/state_machines/base.py]: No need to aggregate the single sig. A sig and an aggregate sig are indistinguishable in BLS

File: eth/beacon/state_machines/validation.py [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+# Parent block proposer validation
+#
+def validate_parent_block_proposer(crystallized_state: 'CrystallizedState',
+                                   block: 'BaseBeaconBlock',
+                                   parent_block: 'BaseBeaconBlock',
+                                   cycle_length: int) -> None:
+    if block.slot_number == 0:
+        return
+
+    block_committees_info = get_block_committees_info(
```
Comment (djrtwo) [inline eth/beacon/state_machines/validation.py]: I really like the BlockCommittee info :)

File: tests/beacon/test_aggregation.py [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+        (0),
+        (9),
+    ],
+)
+def test_aggregate_votes(votes_count, random, privkeys, pubkeys):
+    bit_count = 10
+    pre_bitfield = get_empty_bitfield(bit_count)
+    pre_sigs = ()
+
+    random_votes = random.sample([i for i in range(bit_count)], votes_count)
```
Comment (ChihChengLiang) [inline tests/beacon/test_aggregation.py]: nitpick: could `[i for i in range(bit_count)]` be replaced with ` range(bit_count)`?
```python
>>> import random
>>> random.seed(5566)
>>> random.sample([i for i in range(10)], 5)
[0, 9, 2, 8, 4]
>>> random.seed(5566)
>>> random.sample(range(10), 5)
[0, 9, 2, 8, 4]
```

**Comment by hwwhww** [comment]
@djrtwo @ChihChengLiang thanks for the feedback! ✨ 
I'm going to merge this PR and make other issues for following up attestation aggregation.
